gpio/src/gpio.rs

```diff
@@ -0,0 +1,943 @@
+
+    fn get_num_lines(&self) -> Result<u16>;
+    fn get_line_name(&self, gpio: u16) -> Result<String>;
+    fn get_direction(&self, gpio: u16) -> Result<u8>;
+    fn set_direction(&self, gpio: u16, dir: u8, value: u32) -> Result<()>;
+    fn get_value(&self, gpio: u16) -> Result<u8>;
+    fn set_value(&self, gpio: u16, value: u32) -> Result<()>;
+
+    fn set_irq_type(&self, gpio: u16, value: u16) -> Result<()>;
+    fn wait_for_interrupt(&self, gpio: u16) -> Result<()>;
+}
+
+pub(crate) struct PhysDeviceState {
+    // See wait_for_interrupt() for explanation of Arc.
+    request: Option<Arc<LineRequest>>,
+    buffer: Option<EdgeEventBuffer>,
+}
+
+pub(crate) struct PhysDevice {
+    chip: Chip,
+    ngpio: u16,
+    state: Vec<RwLock<PhysDeviceState>>,
+}
+
+unsafe impl Send for PhysDevice {}
+unsafe impl Sync for PhysDevice {}
+
+impl GpioDevice for PhysDevice {
+    fn open(device: u32) -> Result<Self>
+    where
+        Self: Sized,
+    {
+        let path = format!("/dev/gpiochip{}", device);
+        let chip = Chip::open(&path).map_err(Error::GpiodFailed)?;
+        let ngpio = chip.get_num_lines() as u16;
+
+        // Can't set a vector to all None easily
+        let mut state: Vec<RwLock<PhysDeviceState>> = Vec::new();
+        state.resize_with(ngpio as usize, || {
+            RwLock::new(PhysDeviceState {
+                request: None,
+                buffer: None,
+            })
+        });
+
+        Ok(PhysDevice { chip, ngpio, state })
+    }
+
+    fn get_num_lines(&self) -> Result<u16> {
+        Ok(self.ngpio)
+    }
+
+    fn get_line_name(&self, gpio: u16) -> Result<String> {
+        let line_info = self
+            .chip
+            .line_info(gpio.into())
+            .map_err(Error::GpiodFailed)?;
+
+        Ok(line_info.get_name().unwrap_or("").to_string())
+    }
+
+    fn get_direction(&self, gpio: u16) -> Result<u8> {
+        let line_info = self
+            .chip
+            .line_info(gpio.into())
+            .map_err(Error::GpiodFailed)?;
+
+        Ok(
+            match line_info.get_direction().map_err(Error::GpiodFailed)? {
+                Direction::AsIs => VIRTIO_GPIO_DIRECTION_NONE,
+                Direction::Input => VIRTIO_GPIO_DIRECTION_IN,
+                Direction::Output => VIRTIO_GPIO_DIRECTION_OUT,
+            },
+        )
+    }
+
+    fn set_direction(&self, gpio: u16, dir: u8, value: u32) -> Result<()> {
+        let mut config = LineConfig::new().map_err(Error::GpiodFailed)?;
+        let state = &mut self.state[gpio as usize].write().unwrap();
+
+        match dir {
+            VIRTIO_GPIO_DIRECTION_NONE => {
+                state.request = None;
+                return Ok(());
+            }
+
+            VIRTIO_GPIO_DIRECTION_IN => config.set_direction_offset(Direction::Input, gpio as u32),
+            VIRTIO_GPIO_DIRECTION_OUT => {
+                config.set_direction(Direction::Output);
+                config.set_output_values(&[gpio as u32], &[value as i32]);
```

Comment:
Any reason to use set_output_values() rather than set_output_value(), also provided by the library?   I kept looking for a subtle reason to use slices with a single value rather than just the line and value, but couldn't find one.

Reply:
Right, I can use it.